File: haskell/tests/GoodbyeWorldTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Data.Time.Clock (getCurrentTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Test.Tasty (TestTree, defaultMain, testGroup)
import Test.Tasty.HUnit (assertBool, testCase)

import Goodbye.World (printBanner, printGoodbye)

-- |
-- # @formatErrorMessage@
--
-- The function @formatErrorMessage@ is used to format the error message.
--
-- __Arguments:__
--
-- * @expected: String@ - The expected output.
-- * @actual: String@ - The actual output.
--
-- __Return:__
--
-- * @IO String@ - The formatted error message.
formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

-- |
-- # @printGoodbyeWithoutNameTest@
--
-- The function @printGoodbyeWithoutNameTest@ is used to test the @printGoodbye@
-- function without args.
printGoodbyeWithoutNameTest :: TestTree
printGoodbyeWithoutNameTest = testCase "Print Goodbye without Name Test" $ do
    let args = []
    let expectedOutput = "Goodbye World from Haskell !"
    let actualOutput = printGoodbye args
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

-- |
-- # @printGoodbyeWithNameTest@
--
-- The function @printGoodbyeWithNameTest@ is used to test the @printGoodbye@
-- function with args.
printGoodbyeWithNameTest :: TestTree
printGoodbyeWithNameTest = testCase "Print Goodbye with Name Test" $ do
    let args = ["Alter", "Sabeh"]
    let expectedOutput = "Goodbye Alter Sabeh from Haskell !"
    let actualOutput = printGoodbye args
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

-- |
-- # @printBannerTest@
--
-- The function @printBannerTest@ is used to test the @printBanner@ function.
printBannerTest :: TestTree
printBannerTest = testCase "Print Banner Test" $ do
    let expectedOutput =
            unlines
                [ "                                            "
                , "    ▄████▄  ▄▄▄      ▄▄▄▄   ▄▄▄      ██▓    "
                , "   ▒██▀ ▀█ ▒████▄   ▓█████▄▒████▄   ▓██▒    "
                , "   ▒▓█    ▄▒██  ▀█▄ ▒██▒ ▄█▒██  ▀█▄ ▒██░    "
                , "   ▒▓▓▄ ▄██░██▄▄▄▄██▒██░█▀ ░██▄▄▄▄██▒██░    "
                , "   ▒ ▓███▀ ░▓█   ▓██░▓█  ▀█▓▓█   ▓██░██████▒"
                , "   ░ ░▒ ▒  ░▒▒   ▓▒█░▒▓███▀▒▒▒   ▓▒█░ ▒░▓  ░"
                , "     ░  ▒    ▒   ▒▒ ▒░▒   ░  ▒   ▒▒ ░ ░ ▒  ░"
                , "   ░         ░   ▒   ░    ░  ░   ▒    ░ ░   "
                , "   ░ ░           ░  ░░           ░  ░   ░  ░"
                , "   ░                      ░                 "
                ]
    let actualOutput = printBanner
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

-- |
-- # @tests@
--
-- The @tests@ function contains the tests for the @Goodbye.World@ package.
tests :: TestTree
tests =
    testGroup
        "Goodbye.World Tests"
        [printGoodbyeWithoutNameTest, printGoodbyeWithNameTest, printBannerTest]

-- |
-- # @main@
--
-- The @main@ function is the entrypoint of the test
main :: IO ()
main = defaultMain tests

File: rust/tests/goodbye_world_tests.rs
use chrono::Local;
use colored::Colorize;

use prog::goodbye_world::{print_banner, print_goodbye};

/// # `format_error_message`
///
/// This function formats the error message.
fn format_error_message(expected: &str, actual: &str) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {}\nActual   : {}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

/// # `print_goodbye_without_name_test`
///
/// This function tests the `print_goodbye` function without a name.
#[test]
fn print_goodbye_without_name_test() {
    let args = vec![];
    let expected_output = "Goodbye World from Cargo !";
    let actual_output = print_goodbye(&args);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

/// # `print_goodbye_with_name_test`
///
/// This function tests the `print_goodbye` function with a name.
#[test]
fn print_goodbye_with_name_test() {
    let args = vec!["", "Alter", "Sabeh"];
    let expected_output = "Goodbye Alter Sabeh from Cargo !";
    let actual_output = print_goodbye(&args);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

/// # `print_banner_test`
///
/// This function tests the `print_banner` function.
#[test]
fn print_banner_test() {
    let expected_output = r#"
    ▄████▄  ▄▄▄      ██▀███   ▄████ ▒█████
   ▒██▀ ▀█ ▒████▄   ▓██ ▒ ██▒██▒ ▀█▒██▒  ██▒
   ▒▓█    ▄▒██  ▀█▄ ▓██ ░▄█ ▒██░▄▄▄▒██░  ██▒
   ▒▓▓▄ ▄██░██▄▄▄▄██▒██▀▀█▄ ░▓█  ██▒██   ██░
   ▒ ▓███▀ ░▓█   ▓██░██▓ ▒██░▒▓███▀░ ████▓▒░
   ░ ░▒ ▒  ░▒▒   ▓▒█░ ▒▓ ░▒▓░░▒   ▒░ ▒░▒░▒░
     ░  ▒    ▒   ▒▒ ░ ░▒ ░ ▒░ ░   ░  ░ ▒ ▒░
   ░         ░   ▒    ░░   ░░ ░   ░░ ░ ░ ▒
   ░ ░           ░  ░  ░          ░    ░ ░
   ░
"#
    .to_string();
    let actual_output = print_banner();
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}
